controllers/onload_controller.go

```diff
@@ -78,34 +90,515 @@ func (r *OnloadReconciler) Reconcile(
+		log.Error(err, "Failed to patch module", "Module", module)
+		return nil, err
+	}
+
+	log.Info("Updated module definition for upgrade", "Module", module)
+	return &ctrl.Result{Requeue: true}, nil
+}
+
+const defaultRequeueTime = 5 * time.Second
+
+func (r *OnloadReconciler) handleUpdate(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	res, err := r.handleDevicePluginUpdate(ctx, onload)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	res, err = r.handleModuleUpdate(ctx, onload)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	badNodes, err := r.getBadNodes(ctx, onload)
+	if err != nil {
+		return nil, err
+	}
+	if len(badNodes) == 0 {
+		// Nothing to be done, so return
+		return nil, nil
+	}
+
+	// We just want to upgrade a single node at a time.
+	// Since I don't know if there are any guarantees about how they are ordered
+	// when returned from List(), we shall upgrade nodes by their name
+	// alphabetically.
+	// Golang 1.21 introduces the "slices" package which would allow us to find
+	// the min element of a slice in a single line, however at the moment the
+	// operator is written in Go 1.19 so this package isn't available.
+	// In it's absence I am just using a local funcion.
+
+	node := func(nodes []corev1.Node) corev1.Node {
+		min := nodes[0]
+		for _, node := range nodes {
+			if node.Name < min.Name {
+				min = node
+			}
+		}
+		return min
+	}(badNodes)
+
+	log.Info("Updating onload version on node "+node.Name, "onload", onload)
+
+	// Remove the onload label from the node
+	onloadLabelName := onloadLabelName(onload.Name, onload.Namespace)
+	onloadLabelVersion, found := node.Labels[onloadLabelName]
+	if found && onloadLabelVersion != onload.Spec.Onload.Version {
+		oldNode := node.DeepCopy()
+		delete(node.Labels, onloadLabelName)
+		err := r.Patch(ctx, &node, client.MergeFrom(oldNode))
+		if err != nil {
+			log.Error(err, "Could not patch node to remove onload label",
+				"node", node.Name)
+			return nil, err
+		} else {
+			log.Info("Removed onload label from node " + node.Name)
+			return &ctrl.Result{Requeue: true}, nil
+		}
+	}
+
+	// Check that the device plugin pod has terminated before continuing
+	labelSet := labels.Set{
+		"onload.amd.com/name": onload.Name + devicePluginNameSuffix,
+	}
+	pods, err := r.getPodsOnNode(ctx, labelSet, node.Name)
+	if err != nil {
+		log.Error(err, "Failed to get device plugin pod on node "+node.Name)
+		return nil, err
+	}
+	if len(pods) > 0 {
+		log.Info("Device plugin pod still exists. Re-queueing")
+		return &ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
+	}
+
+	// Evict pods using the onload resource
+	res, err = r.evictOnloadedPods(ctx, node)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	// Remove kmm label
+	oldNode := node.DeepCopy()
+	labelName := kmmLabelName(onload.Name, onload.Namespace)
+	delete(node.Labels, labelName)
+	err = r.Patch(ctx, &node, client.MergeFrom(oldNode))
```

Comment:
Do you think the label deletion (`DeepCopy()`, `delete()`, `Patch()`) deserves a subroutine in the "client_helper" package?